src/cennznut/v0/contract.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright (C) 2019-2020 Centrality Investments Limited
+//!
+//! # CENNZnut - Contracts
+//!
+//! Delegated smart contract permissioning of CENNZnut for use in CENNZnet
+//!
+
+#![warn(clippy::pedantic)]
+
+#[cfg(not(feature = "std"))]
+extern crate alloc;
+
+#[cfg(feature = "std")]
+extern crate std as alloc;
+
+use bit_reverse::ParallelReverse;
+use codec::{Decode, Encode, Input, Output};
+
+/// A CENNZnet permission domain contract
+#[cfg_attr(test, derive(Clone, Debug, Eq, PartialEq))]
+pub struct Contract {
+    pub address: [u8; 32],
+    pub block_cooldown: Option<u32>,
+}
+
+impl Contract {
+    pub fn new(address: &[u8; 32]) -> Self {
+        Self {
+            address: *address,
+            block_cooldown: None,
+        }
+    }
+
+    pub fn wildcard() -> Self {
+        Self {
+            address: [0x00; 32],
+            block_cooldown: None,
+        }
+    }
+
+    pub fn block_cooldown(mut self, block_cooldown: u32) -> Self {
+        self.block_cooldown = Some(block_cooldown);
+        self
+    }
+}
+
+impl Encode for Contract {
+    fn encode_to<T: Output>(&self, buf: &mut T) {
```

Comment:
Have we remained backwards compatible after implementing encode and decode here?